crypto/bn/bn_lib.c

```diff
@@ -739,6 +739,19 @@ void BN_consttime_swap(BN_ULONG condition, BIGNUM *a, BIGNUM *b, int nwords)
     a->top ^= t;
     b->top ^= t;
 
+    t = (a->neg ^ b->neg) & condition;
+    a->neg ^= t;
+    b->neg ^= t;
+
+    /*
+     * cannot just arbitrarily swap flags.
+     * The way a->d is allocated etc.
+     * BN_FLG_MALLOCED, BN_FLG_STATIC_DATA, ...
+     */
+    t = (a->flags ^ b->flags) & condition & BN_FLG_CONSTTIME;
```

Comment:
I'd say that it would be worth explicitly say in commentary that goal it to swap given flag/bit field. I mean it reads "cannot just arbitrarily swap flags ...", and suggestion is "cannot swap all the flags, ..., only single one, BN_FLG_CONSTTIME."

On side note I also wonder if putting it as `((a->flags ^ b->flags) & BN_FLG_CONSTTIME) & condition` would make code more readable. I mean if alternative grouping and explicit parenthesis would illuminate focus on specific flag better... [Side note means that no action is actually expected.]

Reply:
Sure we can modify that comment. It started off as more of a call to core OpenSSL developers -- what flags should we actually be swapping?

Also, defect in BN_swap -- it does not swap BN_FLG_CONSTTIME.

Reply:
Ah! Yeah, that's real danger with looking at suggested changes alone. I mean you risk loosing context. As for flags. Idea behind BN_FLG_STATIC_DATA is actually to indicate that data may not be written to. Intention is actually to treat it as it's read-only data, and some (if not most) of it does reside in read-only segment. In other words observation of BN_FLG_STATIC_DATA in BN_consttime_swap should be treated as fatal condition. It would either cause SEGV or effectively cause data corruption. BN_FLG_MALLOCED refers to BN structure itself, and hence must be preserved. Remaining flags are BN_FLG_CONSTIME and BN_FLG_SECURE. Latter *must* be preserved, because it determines how x->d was allocated and hence how to free it. This leaves BN_FLG_CONSTTIME that one can do something about. To summarize it's sufficient to mask and swap BN_FLG_CONSTTIME alone, i.e. code is correct. Modulo fact that BN_FLG_STATIC_DATA should be treated as fatal...

Reply:
> Modulo fact that BN_FLG_STATIC_DATA should be treated as fatal...

Hmmm... BN_consttime_swap opens with bn_wcheck_size, which actually just a soft assert(*). So that if one of inputs is shorter than intended, it would also cause corruption. In other words this subroutine does make quite strong assumptions about inputs, and it's appropriate to say that absence of BN_FLG_STATIC_DATA is one of such assumptions.

(*) And it's double-conditional, i.e. soft asserts by themselves are conditional on -DNDEBUG, and bn_wcheck_size itself is conditional on BN_DEBUG. One can wonder if BN_consttime_swap deserves a little bit bigger overhaul that would remove this ambiguity, as well as assert for BN_FLG_STATIC_DATA [and remove that switch, it's really no point, loop would do just fine.]

Reply:
+1 we noticed the asserts and no return values. But it feels like a separate PR